[@property] Cache tokens in CSSCustomPropertyValue

anttijk opened the pull request:
#### b63783f818783e6859c645186f4c98a09f4bab2d
<pre>
[@property] Cache tokens in CSSCustomPropertyValue
<a href="https://bugs.webkit.org/show_bug.cgi?id=251672">https://bugs.webkit.org/show_bug.cgi?id=251672</a>
rdar://104997649

Reviewed by Antoine Quint.

We currently generate tokens on every access by tokenizing a (cached) string.

* Source/WebCore/animation/CSSPropertyAnimation.cpp:
(WebCore::blendedCSSCustomPropertyValue):
* Source/WebCore/css/CSSCustomPropertyValue.cpp:
(WebCore::CSSCustomPropertyValue::customCSSText const):
(WebCore::CSSCustomPropertyValue::tokens const):

Add a token cache for registered custom properties.
Also avoid copying by returning a reference.

* Source/WebCore/css/CSSCustomPropertyValue.h:

Also remove an unneeded copy constructor.

Canonical link: <a href="https://commits.webkit.org/259813@main">https://commits.webkit.org/259813@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0a6d0a3cdd1796d1569489e4ec56ec64954f7fb5

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/106040 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/15097 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/38873 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/115223 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/175320 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/109943 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/16521 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/6270 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/98250 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/114948 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/111793 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/12580 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/95551 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/40121 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/94448 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/27193 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/81799 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/8348 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/28545 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/8840 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/5097 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/14461 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/48089 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6791 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/10384 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

graouts commented on Source/WebCore/animation/CSSPropertyAnimation.cpp:
```diff
@@ -3931,7 +3931,7 @@ static Ref<CSSCustomPropertyValue> blendedCSSCustomPropertyValue(const RenderSty
 
     // Use a discrete interpolation for all other cases.
-    return CSSCustomPropertyValue::create(blendingContext.progress < 0.5 ? from : to);
+    return blendingContext.progress < 0.5 ? from : to;
```
Duh, I suppose I wrote that. That is lame. Thanks for fixing.

graouts commented on Source/WebCore/css/CSSCustomPropertyValue.cpp:
```diff
@@ -110,30 +110,38 @@ String CSSCustomPropertyValue::customCSSText() const
-    WTF::switchOn(m_value, [&](const std::monostate&) {
+    static NeverDestroyed<Vector<CSSParserToken>> emptyTokens;
+
+    return WTF::switchOn(m_value, [&](const std::monostate&) -> const Vector<CSSParserToken>& {
```
Are all the explicit `const Vector<CSSParserToken>&` return value types absolutely necessary?

anttijk replied:
yeah, sadly it deduces `Vector<CSSParserToken>` otherwise

anttijk committed:
[@property] Cache tokens in CSSCustomPropertyValue
https://bugs.webkit.org/show_bug.cgi?id=251672
rdar://104997649

Reviewed by Antoine Quint.

We currently generate tokens on every access by tokenizing a (cached) string.

* Source/WebCore/animation/CSSPropertyAnimation.cpp:
(WebCore::blendedCSSCustomPropertyValue):
* Source/WebCore/css/CSSCustomPropertyValue.cpp:
(WebCore::CSSCustomPropertyValue::customCSSText const):
(WebCore::CSSCustomPropertyValue::tokens const):

Add a token cache for registered custom properties.
Also avoid copying by returning a reference.

* Source/WebCore/css/CSSCustomPropertyValue.h:

Also remove an unneeded copy constructor.

Canonical link: https://commits.webkit.org/259813@main
